File: mocks/oppgave-mock/src/main/java/no/nav/pensjon/vtp/mocks/oppgave/repository/OppgaveRepository.java
package no.nav.pensjon.vtp.mocks.oppgave.repository;

import static java.lang.String.format;
import static java.util.Optional.ofNullable;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;

import org.springframework.stereotype.Repository;

@Repository
public class OppgaveRepository {
    private final Map<String, OppgaveFoo> oppgaver = new ConcurrentHashMap<>();

    public Optional<OppgaveFoo> findById(final String oppgaveId) {
        return ofNullable(oppgaver.get(oppgaveId));
    }

    public Stream<OppgaveFoo> findAll() {
        return oppgaver.values().stream();
    }

    public String saveOppgave(final OppgaveFoo oppgaveFoo) {
        ofNullable(oppgaver.get(oppgaveFoo.getOppgaveId()))
                .filter(oppgaveFoo::isDifferentVersion)
                .ifPresent(existingOppgaveFoo -> {
                    throw new IllegalArgumentException(
                            format("Oppgaven kan ikke oppdateres når gjeldende versjon og innsendt versjon ikke er like. Gjeldende versjon: %d, angitt versjon: %d uuid: %s",
                                    existingOppgaveFoo.getVersion(), oppgaveFoo.getVersion(), oppgaveFoo.getOppgaveId())
                    );
                });

        this.oppgaver.put(oppgaveFoo.getOppgaveId(), oppgaveFoo.withIncrementedVersion());
        return oppgaveFoo.getOppgaveId();
    }

}

File: mocks/oppgave-mock/src/main/java/no/nav/pensjon/vtp/mocks/oppgave/repository/OppgaveRepository.kt
package no.nav.pensjon.vtp.mocks.oppgave.repository

import no.nav.pensjon.vtp.core.util.toOptional
import org.springframework.stereotype.Repository
import java.util.*
import java.util.concurrent.ConcurrentHashMap

@Repository
class OppgaveRepository : OppgaveFooBaseRepository {
    private val oppgaver: MutableMap<String, OppgaveFoo> = ConcurrentHashMap()

    override fun findById(id: String): Optional<OppgaveFoo> = oppgaver[id].toOptional()

    override fun findAll() : Collection<OppgaveFoo> = oppgaver.values

    override fun save(oppgave: OppgaveFoo): OppgaveFoo {
        oppgaver[oppgave.oppgaveId]
                ?.takeIf(oppgave::hasDifferentVersion)
                ?.let {
                    throw IllegalArgumentException(
                            "Oppgaven kan ikke oppdateres når gjeldende versjon og innsendt versjon ikke er like. " +
                                    "Gjeldende versjon: ${it.version}, angitt versjon: ${oppgave.version} uuid: ${oppgave.oppgaveId}"
                    )
                }

        oppgaver[oppgave.oppgaveId] = oppgave.withIncrementedVersion()
        return oppgave
    }
}
